internal/database/testing/database.go

```diff
@@ -51,3 +52,60 @@ func DumpTable(c *gc.C, db *sql.DB, table string) {
 	writer.Flush()
 	fmt.Fprintln(os.Stdout)
 }
+
+// DumpTablesJSON prints the named tables to stdout in a JSON format. This is
+// intended as a debugging tool for state tests. It is not intended for use in
+// production code.
+func DumpTablesJSON(c *gc.C, db *sql.DB, tableNames ...string) {
+	for _, tableName := range tableNames {
+		query := fmt.Sprintf("SELECT * FROM %s", tableName)
+		table := &tableData{
+			tableName: tableName,
+		}
+
+		rows, err := db.Query(query)
+		c.Assert(err, jc.ErrorIsNil)
+
+		columns, err := rows.Columns()
+		c.Assert(err, jc.ErrorIsNil)
+		table.columnNames = columns
+
+		for rows.Next() {
+			row := make([]any, len(columns))
+			for i := range row {
+				row[i] = new([]byte)
+			}
+			err = rows.Scan(row...)
+			c.Assert(err, jc.ErrorIsNil)
+			table.rows = append(table.rows, row)
+		}
+
+		printJSON(c, table)
+	}
+
+}
+
+// tableData is an intermediate representation of an arbitrary table, ready to
+// be processed for printing.
+type tableData struct {
+	tableName   string
+	columnNames []string
+	rows        [][]any
+}
+
+func printJSON(c *gc.C, table *tableData) {
+	fmt.Printf("***** TABLE %q *****\n", table.tableName)
+	stdout := json.NewEncoder(os.Stdout)
+
+	for _, row := range table.rows {
+		jsonRow := map[string]string{}
+		for i, columnName := range table.columnNames {
+			jsonRow[columnName] = fmt.Sprintf("%s", *row[i].(*[]byte))
```

Comment:
This is pretty reckless just taking a random pointer to some data and turning it into a bag of bytes for a string. I don't think this is going to work how you expect it an also produce some pretty invalid json. Why not just let the json encoder assume the type and output the correct json type for the val?

Reply:
Good point